internal/expr/typeinfo.go

```diff
@@ -109,9 +112,59 @@ func getTypeInfo(value any) (typeInfo, error) {
 	return typeInfo, nil
 }
 
+// getStructFields requires the caller to check that t is a struct.
+func getStructFields(t reflect.Type) ([]*structField, error) {
+	var fields []*structField
+	for i := 0; i < t.NumField(); i++ {
+		f := t.Field(i)
+		if !f.IsExported() {
+			return nil, fmt.Errorf("field %q of struct %s not exported", f.Name, t.Name())
+		}
+		ft := f.Type
+		k := ft.Kind()
+		// Check if the field type is an embedded/nested struct or pointer to
+		// one. If a pointer to the field type implementes Scanner then it is
+		// not a nested struct.
+		if (k == reflect.Struct && !reflect.PointerTo(ft).Implements(scannerInterface)) ||
+			(k == reflect.Pointer && ft.Elem().Kind() == reflect.Struct && !ft.Implements(scannerInterface)) {
+			if k == reflect.Pointer {
+				ft = ft.Elem()
+			}
+			nestedFields, err := getStructFields(ft)
+			if err != nil {
+				return nil, err
+			}
+			for _, nestedField := range nestedFields {
+				nestedField.index = append([]int{i}, nestedField.index...)
+				nestedField.structType = t
+			}
+			fields = append(fields, nestedFields...)
+		} else {
+			// Fields without a "db" tag are outside of SQLair's remit.
+			tag := f.Tag.Get("db")
+			if tag == "" {
+				continue
+			}
+			tag, omitEmpty, err := parseTag(tag)
+			if err != nil {
+				return nil, fmt.Errorf("cannot parse tag for field %s.%s: %s", t.Name(), f.Name, err)
+			}
+			fields = append(fields, &structField{
+				name:       f.Name,
+				index:      f.Index,
```

Comment:
If I understand the code correctly, reflect.StructField.Index will only ever contain a single element, in other words, it contains only an index within its immediate parent struct (even though the parent struct may be embedded in another struct)? If this is true, it makes sense why we have to prepend the parent struct Index each time a recursive iteration completes.